core/state/database.go

```diff
@@ -253,36 +254,92 @@ func (tds *TrieDbState) Copy() *TrieDbState {
 		incarnationMap: make(map[common.Hash]uint64),
 	}
 
-	if !tds.noIntermediateCache {
-		cpy.t.SetUnloadFunc(cpy.putIntermediateCache)
-		cpy.tp.SetCreateNodeFunc(cpy.delIntermediateCache)
+	if debug.IsIntermediateTrieHash() {
+		cpy.t.SetUnloadNodeFunc(cpy.putIntermediateHash)
+		cpy.t.SetOnDeleteSubtreeFunc(tds.markSubtreeEmptyInIntermediateHash)
+		cpy.tp.SetCreateNodeFunc(cpy.delIntermediateHash)
 	}
 
 	return &cpy
 }
 
-func (tds *TrieDbState) putIntermediateCache(prefix []byte, nodeHash []byte) {
-	if len(prefix)%2 == 1 { // only put to bucket prefixes with even number of nibbles
+func (tds *TrieDbState) markSubtreeEmptyInIntermediateHash(prefix []byte) {
+	if len(prefix) != common.HashLength {
 		return
 	}
-	if len(prefix) == 0 {
+
+	db := tds.db
+	if hasDb, ok := tds.db.(ethdb.HasDb); ok {
+		db = hasDb.DB()
+	}
+
+	if err := db.Put(dbutils.IntermediateTrieHashBucket, prefix, []byte{}); err != nil {
+		log.Warn("could not put intermediate trie hash", "err", err)
+	}
+}
+
+func (tds *TrieDbState) putIntermediateHash(prefixAsNibbles []byte, nodeHash []byte) {
+	if len(prefixAsNibbles) == 0 {
 		return
 	}
-	if err := putIntermediateCache(tds.db, prefix, nodeHash); err != nil {
-		log.Warn("could not put intermediate trie cache", "err", err)
+
+	if len(prefixAsNibbles)%2 == 1 { // only put to bucket prefixes with even number of prefixAsNibbles
+		return
+	}
+
+	key := pool.GetBuffer(64)
+	defer pool.PutBuffer(key)
+
+	if err := trie.CompressNibbles(prefixAsNibbles, &key.B); err != nil {
+		log.Warn("could not CompressNibbles for intermediate trie hash", "err", err)
+		return
+	}
+
+	v := common.CopyBytes(nodeHash)
+
+	db := tds.db
+	if hasDb, ok := tds.db.(ethdb.HasDb); ok {
+		db = hasDb.DB()
+	} else {
+		fmt.Printf("what is it? %T\n", tds.db)
 	}
 
+	//	fmt.Printf("Put: %x %x\n", key.Bytes(), v)
+	if err := db.Put(dbutils.IntermediateTrieHashBucket, key.Bytes(), v); err != nil {
```

Comment:
So you are going directly to the underlying database here? I suspect for two reasons. First, the `key` does not always survive until the commit of the mutation (this can be fixed by simply making a copy instead of using pool buffer). Second, you might be using a walk somewhere. But this walk does not need to see uncommitted data anyway. So I suggest you remove this and simply make the copy of the key above, instead of using the buffer

Reply:
I did as you said, it works, but I don't know why :-) Why "this walk does not need to see uncommitted data anyway"?